esx_service/utils/vm_listener.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python
+# Copyright 2016 VMware, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+'''
+VM change listener (started as a part of vmdkops service).
+It monitors VM poweroff events and detaches the DVS managed
+volumes from the VM and updates the status in KV
+'''
+
+import logging
+import os
+import os.path
+import atexit
+
+import threadutils
+import log_config
+import vmdk_utils
+import vmdk_ops
+
+from pyVmomi import VmomiSupport, vim, vmodl
+
+VM_POWERSTATE = 'runtime.powerState'
+POWERSTATE_POWEROFF = 'poweredOff'
+
+
+def start_vm_changelistener():
+    """
+    Listen to power state changes of VMs running on current host
+    """
+    si = vmdk_ops.get_si()
+    pc = si.content.propertyCollector
+    err_msg = create_vm_powerstate_filter(pc, si.content.rootFolder)
+    if err_msg:
+        return err_msg
+    # Start a separate thread to listen to changes
+    threadutils.start_new_thread(target=listen_vm_propertychange,
+                                 args=(pc,),
+                                 daemon=True)
+
+
+def create_vm_powerstate_filter(pc, from_node):
+    """
+    Create a filter spec to list to VM power state changes
+    """
+
+    filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+    objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=from_node,
+                                                       selectSet=vm_folder_traversal())
+    filterSpec.objectSet.append(objSpec)
+    # Add the property specs
+    propSpec = vmodl.query.PropertyCollector.PropertySpec(type=vim.VirtualMachine, all=False)
+    propSpec.pathSet.append(VM_POWERSTATE)
+    filterSpec.propSet.append(propSpec)
+    try:
+        pcFilter = pc.CreateFilter(filterSpec, True)
+        atexit.register(pcFilter.Destroy)
+        return None
+    except Exception as e:
+        err_msg = "Problem creating PropertyCollector filter: {}".format(str(e))
+        logging.error(err_msg)
+        return err_msg
+
+
+def listen_vm_propertychange(pc):
+    """
+    Waits for updates on powerstate of VMs. If powerstate is poweroff,
+    detach the dvs managed volumes attached to VM
+    """
+    logging.info("VMChangeListener thread started")
+    version = ''
+    while True:
+        result = pc.WaitForUpdates(version)
```

Comment:
1. Stopping service stops/exits Listerner thread right?
2. Make sure this thread doesn't hog CPU. check esxtop